File: internal/appui/input_mentions.go
package appui

import (
	"encoding/json"
	"net/url"
	"strings"

	"aiops-v2/internal/hostops"
)

const metadataInputMentionsV1 = "aiops.input.mentions.v1"

type inputMentionEnvelope struct {
	Version  int                   `json:"version"`
	Mentions []inputMentionBinding `json:"mentions"`
}

type inputMentionBinding struct {
	Version int                    `json:"version"`
	TokenID string                 `json:"tokenId"`
	Sigil   string                 `json:"sigil"`
	Display string                 `json:"display"`
	RawText string                 `json:"rawText"`
	Kind    string                 `json:"kind"`
	Path    string                 `json:"path"`
	Source  string                 `json:"source"`
	Range   inputMentionRange      `json:"range"`
	Payload map[string]interface{} `json:"payload"`
}

type inputMentionRange struct {
	Start int `json:"start"`
	End   int `json:"end"`
}

type inputMentionHostHint struct {
	TokenID     string
	Raw         string
	SpanStart   int
	SpanEnd     int
	HostID      string
	Address     string
	DisplayName string
	Source      string
}

type inputMentionResourceHint struct {
	TokenID string
	Raw     string
	Kind    string
	ID      string
	Title   string
}

type parsedInputMentions struct {
	Present      bool
	Invalid      bool
	Source       string
	Validation   string
	Hosts        []inputMentionHostHint
	Capabilities []string
	Resources    []inputMentionResourceHint
}

func (p parsedInputMentions) HasCapability(name string) bool {
	name = strings.TrimSpace(strings.ToLower(name))
	for _, capability := range p.Capabilities {
		if capability == name {
			return true
		}
	}
	return false
}

func parseInputMentions(input string, metadata map[string]string) parsedInputMentions {
	raw := strings.TrimSpace(metadata[metadataInputMentionsV1])
	if raw == "" {
		return parsedInputMentions{Source: "absent", Validation: "absent"}
	}
	parsed := parsedInputMentions{Present: true, Source: "structured", Validation: "confirmed"}
	var envelope inputMentionEnvelope
	if err := json.Unmarshal([]byte(raw), &envelope); err != nil || envelope.Version != 1 {
		parsed.Invalid = true
		parsed.Validation = "invalid"
		return parsed
	}
	for _, mention := range envelope.Mentions {
		startByte, endByte, ok := inputMentionTextMatchRange(input, mention)
		if !ok {
			parsed.Invalid = true
			continue
		}
		if !isStrongInputMentionSource(mention.Source) {
			if isWeakInputMentionSource(mention.Source) {
				continue
			}
			parsed.Invalid = true
			continue
		}
		switch strings.ToLower(strings.TrimSpace(mention.Kind)) {
		case "host":
			hostID, ok := decodeHostMentionPath(mention.Path)
			if !ok {
				parsed.Invalid = true
				continue
			}
			if payloadHostID := stringPayload(mention.Payload, "hostId"); payloadHostID != "" {
				hostID = payloadHostID
			}
			parsed.Hosts = append(parsed.Hosts, inputMentionHostHint{
				TokenID:     strings.TrimSpace(mention.TokenID),
				Raw:         firstNonEmptyString(strings.TrimSpace(mention.RawText), strings.TrimSpace(mention.Display)),
				SpanStart:   startByte,
				SpanEnd:     endByte,
				HostID:      hostID,
				Address:     firstNonEmptyString(stringPayload(mention.Payload, "address"), hostID),
				DisplayName: firstNonEmptyString(stringPayload(mention.Payload, "displayName"), hostID),
				Source:      strings.TrimSpace(mention.Source),
			})
		case "capability":
			capability, ok := decodeCapabilityMentionPath(mention.Path)
			if !ok {
				parsed.Invalid = true
				continue
			}
			parsed.Capabilities = appendUniqueInputMentionString(parsed.Capabilities, capability)
		case "ops_manual":
			manualID, ok := decodeResourceMentionPath(mention.Path, "ops-manual://")
			if !ok {
				parsed.Invalid = true
				continue
			}
			if payloadManualID := stringPayload(mention.Payload, "manualId"); payloadManualID != "" {
				manualID = payloadManualID
			}
			parsed.Capabilities = appendUniqueInputMentionString(parsed.Capabilities, "ops_manuals")
			parsed.Resources = append(parsed.Resources, inputMentionResourceHint{
				TokenID: strings.TrimSpace(mention.TokenID),
				Raw:     firstNonEmptyString(strings.TrimSpace(mention.RawText), strings.TrimSpace(mention.Display)),
				Kind:    "ops_manual",
				ID:      manualID,
				Title:   firstNonEmptyString(stringPayload(mention.Payload, "title"), manualID),
			})
		case "ops_graph":
			graphID, ok := decodeResourceMentionPath(mention.Path, "ops-graph://")
			if !ok {
				parsed.Invalid = true
				continue
			}
			if payloadGraphID := stringPayload(mention.Payload, "graphId"); payloadGraphID != "" {
				graphID = payloadGraphID
			}
			parsed.Capabilities = appendUniqueInputMentionString(parsed.Capabilities, "ops_graph")
			parsed.Resources = append(parsed.Resources, inputMentionResourceHint{
				TokenID: strings.TrimSpace(mention.TokenID),
				Raw:     firstNonEmptyString(strings.TrimSpace(mention.RawText), strings.TrimSpace(mention.Display)),
				Kind:    "ops_graph",
				ID:      graphID,
				Title:   firstNonEmptyString(stringPayload(mention.Payload, "name"), graphID),
			})
		default:
			parsed.Invalid = true
		}
	}
	if parsed.Invalid {
		parsed.Validation = "invalid"
	}
	return parsed
}

func inputMentionTextMatchRange(input string, mention inputMentionBinding) (int, int, bool) {
	if mention.Version != 1 || mention.Sigil != "@" {
		return 0, 0, false
	}
	startByte, endByte, ok := inputMentionUTF16RangeToByteOffsets(input, mention.Range.Start, mention.Range.End)
	if !ok {
		return 0, 0, false
	}
	return startByte, endByte, input[startByte:endByte] == mention.RawText
}

func inputMentionHostHintsToHostMentions(hints []inputMentionHostHint) []hostops.HostMention {
	mentions := make([]hostops.HostMention, 0, len(hints))
	for _, hint := range hints {
		mentions = append(mentions, hostops.HostMention{
			TokenID:     strings.TrimSpace(hint.TokenID),
			Raw:         strings.TrimSpace(hint.Raw),
			SpanStart:   hint.SpanStart,
			SpanEnd:     hint.SpanEnd,
			HostID:      strings.TrimSpace(hint.HostID),
			Address:     strings.TrimSpace(hint.Address),
			DisplayName: strings.TrimSpace(hint.DisplayName),
			Source:      hostops.HostMentionSourceInventory,
			Resolved:    false,
			Confidence:  0.75,
		})
	}
	return mentions
}

func inputMentionUTF16RangeToByteOffsets(input string, start, end int) (int, int, bool) {
	if start < 0 || end <= start {
		return 0, 0, false
	}
	unitIndex := 0
	startByte := -1
	endByte := -1
	for byteIndex, r := range input {
		if unitIndex == start {
			startByte = byteIndex
		}
		if unitIndex == end {
			endByte = byteIndex
			break
		}
		unitIndex += inputMentionUTF16Width(r)
		if startByte < 0 && unitIndex == start {
			startByte = byteIndex + len(string(r))
		}
		if unitIndex == end {
			endByte = byteIndex + len(string(r))
			break
		}
		if unitIndex > end {
			return 0, 0, false
		}
	}
	if startByte < 0 && unitIndex == start {
		startByte = len(input)
	}
	if endByte < 0 && unitIndex == end {
		endByte = len(input)
	}
	if startByte < 0 || endByte < 0 || endByte <= startByte {
		return 0, 0, false
	}
	return startByte, endByte, true
}

func inputMentionUTF16Width(r rune) int {
	if r > 0xFFFF {
		return 2
	}
	return 1
}

func decodeHostMentionPath(path string) (string, bool) {
	raw, ok := strings.CutPrefix(strings.TrimSpace(path), "host://")
	if !ok || raw == "" {
		return "", false
	}
	decoded, err := url.PathUnescape(raw)
	if err != nil {
		return raw, true
	}
	return strings.TrimSpace(decoded), true
}

func decodeCapabilityMentionPath(path string) (string, bool) {
	raw, ok := strings.CutPrefix(strings.TrimSpace(path), "capability://")
	if !ok {
		return "", false
	}
	switch strings.ToLower(strings.TrimSpace(raw)) {
	case "coroot":
		return "coroot", true
	case "ops_graph":
		return "ops_graph", true
	case "ops_manuals":
		return "ops_manuals", true
	default:
		return "", false
	}
}

func decodeResourceMentionPath(path, prefix string) (string, bool) {
	raw, ok := strings.CutPrefix(strings.TrimSpace(path), prefix)
	if !ok || raw == "" {
		return "", false
	}
	decoded, err := url.PathUnescape(raw)
	if err != nil {
		return raw, true
	}
	return strings.TrimSpace(decoded), true
}

func stringPayload(payload map[string]interface{}, key string) string {
	if payload == nil {
		return ""
	}
	value, ok := payload[key]
	if !ok {
		return ""
	}
	text, ok := value.(string)
	if !ok {
		return ""
	}
	return strings.TrimSpace(text)
}

func isStrongInputMentionSource(source string) bool {
	switch strings.ToLower(strings.TrimSpace(source)) {
	case "selection", "history_restore":
		return true
	default:
		return false
	}
}

func isWeakInputMentionSource(source string) bool {
	return strings.EqualFold(strings.TrimSpace(source), "typed_fallback")
}

func appendUniqueInputMentionString(values []string, next string) []string {
	next = strings.TrimSpace(strings.ToLower(next))
	if next == "" {
		return values
	}
	for _, value := range values {
		if value == next {
			return values
		}
	}
	return append(values, next)
}

File: web/src/chat/inputMentions.ts
export const INPUT_MENTIONS_METADATA_KEY = "aiops.input.mentions.v1";

export type AiopsMentionKind =
  | "host"
  | "capability"
  | "ops_manual"
  | "ops_graph"
  | "file"
  | "unknown";

export type AiopsMentionSource =
  | "selection"
  | "typed_fallback"
  | "history_restore";

export type AiopsMentionRange = {
  start: number;
  end: number;
};

export type AiopsMentionBinding = {
  version: 1;
  tokenId: string;
  sigil: "@";
  display: string;
  rawText: string;
  kind: AiopsMentionKind;
  path: string;
  source: AiopsMentionSource;
  range: AiopsMentionRange;
  selectedAtMs?: number;
  payload?: Record<string, unknown>;
};

export type HostMentionBindingInput = {
  tokenId: string;
  rawText: string;
  range: AiopsMentionRange;
  hostId: string;
  address?: string;
  displayName?: string;
  status?: string;
  source?: AiopsMentionSource;
};

export type CapabilityMention = "coroot" | "ops_graph" | "ops_manuals";

export type CapabilityMentionBindingInput = {
  tokenId: string;
  rawText: string;
  range: AiopsMentionRange;
  capability: CapabilityMention;
  source?: AiopsMentionSource;
};

export type OpsManualMentionBindingInput = {
  tokenId: string;
  rawText: string;
  range: AiopsMentionRange;
  manualId: string;
  title: string;
  workflowId?: string;
  status?: string;
  source?: AiopsMentionSource;
};

export type OpsGraphMentionBindingInput = {
  tokenId: string;
  rawText: string;
  range: AiopsMentionRange;
  graphId: string;
  name: string;
  environment?: string;
  source?: AiopsMentionSource;
};

type InputMentionEnvelope = {
  version: 1;
  mentions: AiopsMentionBinding[];
};

export function buildHostMentionBinding(input: HostMentionBindingInput): AiopsMentionBinding {
  const hostId = cleanText(input.hostId);
  const rawText = cleanText(input.rawText);
  return {
    version: 1,
    tokenId: cleanText(input.tokenId),
    sigil: "@",
    display: rawText,
    rawText,
    kind: "host",
    path: `host://${encodeURIComponent(hostId)}`,
    source: input.source || "selection",
    range: normalizeRange(input.range),
    selectedAtMs: Date.now(),
    payload: {
      hostId,
      address: cleanText(input.address) || hostId,
      displayName: cleanText(input.displayName) || hostId,
      ...(cleanText(input.status) ? { status: cleanText(input.status) } : {}),
    },
  };
}

export function buildCapabilityMentionBinding(input: CapabilityMentionBindingInput): AiopsMentionBinding {
  const capability = normalizeCapability(input.capability);
  const rawText = cleanText(input.rawText);
  return {
    version: 1,
    tokenId: cleanText(input.tokenId),
    sigil: "@",
    display: rawText,
    rawText,
    kind: "capability",
    path: `capability://${capability}`,
    source: input.source || "selection",
    range: normalizeRange(input.range),
    selectedAtMs: Date.now(),
  };
}

export function buildOpsManualMentionBinding(input: OpsManualMentionBindingInput): AiopsMentionBinding {
  const manualId = cleanText(input.manualId);
  const rawText = cleanText(input.rawText);
  const title = cleanText(input.title) || manualId;
  const workflowId = cleanText(input.workflowId);
  const status = cleanText(input.status);
  return {
    version: 1,
    tokenId: cleanText(input.tokenId),
    sigil: "@",
    display: title || rawText,
    rawText,
    kind: "ops_manual",
    path: `ops-manual://${encodeURIComponent(manualId)}`,
    source: input.source || "selection",
    range: normalizeRange(input.range),
    selectedAtMs: Date.now(),
    payload: {
      manualId,
      title,
      ...(workflowId ? { workflowId } : {}),
      ...(status ? { status } : {}),
    },
  };
}

export function buildOpsGraphMentionBinding(input: OpsGraphMentionBindingInput): AiopsMentionBinding {
  const graphId = cleanText(input.graphId);
  const rawText = cleanText(input.rawText);
  const name = cleanText(input.name) || graphId;
  const environment = cleanText(input.environment);
  return {
    version: 1,
    tokenId: cleanText(input.tokenId),
    sigil: "@",
    display: name || rawText,
    rawText,
    kind: "ops_graph",
    path: `ops-graph://${encodeURIComponent(graphId)}`,
    source: input.source || "selection",
    range: normalizeRange(input.range),
    selectedAtMs: Date.now(),
    payload: {
      graphId,
      name,
      ...(environment ? { environment } : {}),
    },
  };
}

export function reconcileMentionBindings(
  text: string,
  bindings: AiopsMentionBinding[],
): AiopsMentionBinding[] {
  const currentText = String(text || "");
  return bindings.filter((binding) => {
    if (!binding || binding.version !== 1) return false;
    const range = normalizeRange(binding.range);
    if (range.start < 0 || range.end > currentText.length || range.end <= range.start) {
      return false;
    }
    return currentText.slice(range.start, range.end) === binding.rawText;
  });
}

export function buildInputMentionMetadata(bindings: AiopsMentionBinding[]): Record<string, string> {
  const normalized = bindings.filter((binding) => binding && binding.version === 1);
  if (normalized.length === 0) return {};
  const envelope: InputMentionEnvelope = { version: 1, mentions: normalized };
  return { [INPUT_MENTIONS_METADATA_KEY]: JSON.stringify(envelope) };
}

export function deriveHostMentionMetadata(bindings: AiopsMentionBinding[]): Record<string, string> {
  const hostMentions = bindings
    .filter((binding) => binding.kind === "host" && isStrongMentionSource(binding.source))
    .map((binding) => {
      const payload = objectPayload(binding.payload);
      const hostId = cleanText(payload.hostId) || decodeHostPath(binding.path);
      const address = cleanText(payload.address) || hostId;
      const displayName = cleanText(payload.displayName) || hostId;
      return {
        tokenId: binding.tokenId,
        raw: binding.rawText,
        value: hostId,
        start: binding.range.start,
        end: binding.range.end,
        hostId,
        address,
        displayName,
        source: "inventory",
        resolved: true,
        confidence: 1,
      };
    })
    .filter((mention) => mention.hostId);

  if (hostMentions.length === 0) return {};
  return {
    "aiops.hostops.mentions": JSON.stringify(hostMentions),
    "aiops.hostops.clientDetectedMultiHost": String(uniqueHostKeys(hostMentions).length >= 2),
  };
}

export function deriveCapabilityMentionMetadata(bindings: AiopsMentionBinding[]): Record<string, string> {
  const packs: string[] = [];
  const tools: string[] = [];
  const metadata: Record<string, string> = {};
  for (const binding of bindings) {
    if (!isStrongMentionSource(binding.source)) continue;
    if (binding.kind === "ops_graph") {
      const payload = objectPayload(binding.payload);
      const graphId = cleanText(payload.graphId) || decodeResourcePath(binding.path, "ops-graph://");
      const name = cleanText(payload.name) || graphId;
      packs.push("opsgraph");
      metadata["aiops.opsGraph.explicitMention"] = "true";
      if (graphId) metadata["aiops.opsGraph.graphId"] = graphId;
      if (name) metadata["aiops.opsGraph.graphName"] = name;
      continue;
    }
    if (binding.kind === "ops_manual") {
      const payload = objectPayload(binding.payload);
      const manualId = cleanText(payload.manualId) || decodeResourcePath(binding.path, "ops-manual://");
      const workflowId = cleanText(payload.workflowId);
      const title = cleanText(payload.title) || manualId;
      packs.push("ops_manual_flow");
      tools.push("search_ops_manuals");
      metadata["aiops.opsManuals.explicitMention"] = "true";
      if (manualId) {
        metadata.opsManualManualId = manualId;
        metadata.manualId = manualId;
      }
      if (workflowId) {
        metadata.opsManualWorkflowId = workflowId;
        metadata.workflowId = workflowId;
      }
      if (title) metadata.opsManualManualTitle = title;
      continue;
    }
    if (binding.kind !== "capability") continue;
    const capability = decodeCapabilityPath(binding.path);
    if (capability === "coroot") {
      metadata["aiops.coroot.explicitRCA"] = "true";
      metadata["aiops.coroot.rcaDisplayAllowed"] = "true";
    }
    if (capability === "ops_graph") {
      packs.push("opsgraph");
      metadata["aiops.opsGraph.explicitMention"] = "true";
    }
    if (capability === "ops_manuals") {
      packs.push("ops_manual_flow");
      tools.push("search_ops_manuals");
      metadata["aiops.opsManuals.explicitMention"] = "true";
    }
  }
  if (packs.length > 0) metadata.enableToolPack = uniqueValues(packs).join(",");
  if (tools.length > 0) metadata.enableTool = uniqueValues(tools).join(",");
  return metadata;
}

function normalizeRange(range: AiopsMentionRange): AiopsMentionRange {
  const start = Number.isFinite(range?.start) ? Math.max(0, Math.trunc(range.start)) : 0;
  const end = Number.isFinite(range?.end) ? Math.max(start, Math.trunc(range.end)) : start;
  return { start, end };
}

function normalizeCapability(value: CapabilityMention): CapabilityMention {
  if (value === "coroot" || value === "ops_graph" || value === "ops_manuals") {
    return value;
  }
  return "coroot";
}

function decodeHostPath(path: string) {
  const raw = cleanText(path).replace(/^host:\/\//, "");
  try {
    return decodeURIComponent(raw);
  } catch {
    return raw;
  }
}

function decodeCapabilityPath(path: string): CapabilityMention | "" {
  const raw = cleanText(path).replace(/^capability:\/\//, "");
  if (raw === "coroot" || raw === "ops_graph" || raw === "ops_manuals") return raw;
  return "";
}

function decodeResourcePath(path: string, prefix: string) {
  const raw = cleanText(path).replace(prefix, "");
  try {
    return decodeURIComponent(raw);
  } catch {
    return raw;
  }
}

function objectPayload(value: unknown): Record<string, unknown> {
  return value && typeof value === "object" && !Array.isArray(value)
    ? (value as Record<string, unknown>)
    : {};
}

function uniqueHostKeys(items: Array<{ hostId: string; address: string }>) {
  return uniqueValues(items.map((item) => cleanText(item.hostId) || cleanText(item.address)).filter(Boolean));
}

function uniqueValues(values: string[]) {
  return Array.from(new Set(values.map(cleanText).filter(Boolean)));
}

function isStrongMentionSource(source: AiopsMentionSource) {
  return source === "selection" || source === "history_restore";
}

function cleanText(value: unknown) {
  return typeof value === "string" ? value.trim() : "";
}
